### PR TITLE
Major ai coverity fixes

### PR DESCRIPTION
**Pull request by dnk777**
Not sure if `memset()` counts as fields initialization for the analyzer. Also I'm not sure if the logic is not broken after these changes (I had already fixed a few new glitches introduced by these changes I have managed to find). There is a separate commit fixing the most concerning glitch due to its not obvious nature. The bot behavior looks fine now, and also the bot code is a subject of pending overview and improvement anyway. I have tested gameplay a bit running executables built with AddressSanitizer turned on.

### CODE REVIEW

**Review by viciious — COMMENTED** [review]
Don't you think this is kinda hacky, ugly and can be done in a better way?

**Comment by dnk777** [comment]
Well wait for a rewritten commit. I will remove these filelds from `PoolItem` and will use headers in chunks visible to `PoolBase` only.

**Review by viciious — COMMENTED** [review]
What does this code do and why does it have to be so complex anyway?

File: source/game/ai/ai_base_brain.cpp [hunk 1]
```diff
@@ -63,35 +63,39 @@ void *PoolBase::Alloc() {
...
-	ItemAt( 0 ).nextInList = 1;
-	ItemAt( lastIndex ).prevInList = (short)( lastIndex - 1 );
-	ItemAt( lastIndex ).nextInList = -1;
-	for( short i = 1; i < lastIndex; ++i ) {
-		ItemAt( i ).nextInList = (short)( i + 1 );
-		ItemAt( i ).prevInList = (short)( i - 1 );
+	int16_t lastIndex = (int16_t)( itemsCount - 1 );
+	ItemLinksAt( 0 ).Prev() = -1;
+	ItemLinksAt( 0 ).Next() = 1;
+	ItemLinksAt( lastIndex ).Prev() = (short)( lastIndex - 1 );
```
Comment (viciious) [inline source/game/ai/ai_base_brain.cpp]: Use int16_t instead of short.

**Comment by dnk777** [comment]
Its an implementation of a reusable pool of items. An item can also delete itself from the pool, thats why it has a reference to a parent pool. It must be non-generic (use virtual calls instead of templates) because a GOAP `PlannerNode` operates on non-generic subclasses of an `ActionRecord`. If the planned nodes graph built during the GOAP planning algorithm execution is being cleaned all action records of different kinds (allocated by different GOAP actions) contained in the nodes must be released. There are also generic subtypes of `PoolBase` that allow type-safe allocation of action records.

See https://github.com/Qfusion/qfusion/blob/master/source/game/ai/bot_actions.cpp for examples of its usage (`pool.New(...` calls).

**Comment by dnk777** [comment]
Also the planner node is a pooled item too.

**Comment by dnk777** [comment]
And also not all planner nodes (and associated action records) are removed after planning, ones that are in the built «path» in the logical world state states graph (in the build actions plan) are kept until replanning. Thats why linked lists of free nodes of different kind must be used, and stack-based allocators do not fit this case.

**Pull request merged**